src/leap/mail/adaptors/soledad.py

```diff
@@ -0,0 +1,1047 @@
+        mbox_uuid = mbox_uuid.replace('-', '_')
+        chash = re.findall(constants.FDOCID_CHASH_RE, self.fdoc)[0]
+        new_id = constants.METAMSGID.format(mbox_uuid=mbox_uuid, chash=chash)
+        new_fdoc_id = constants.FDOCID.format(mbox_uuid=mbox_uuid, chash=chash)
+        self._future_doc_id = new_id
+        self.fdoc = new_fdoc_id
+
+
+class MessageWrapper(object):
+
+    # This could benefit of a DeferredLock to create/update all the
+    # documents at the same time maybe, and defend against concurrent updates?
+
+    implements(IMessageWrapper)
+
+    def __init__(self, mdoc, fdoc, hdoc, cdocs=None):
+        """
+        Need at least a metamsg-document, a flag-document and a header-document
+        to instantiate a MessageWrapper. Content-documents can be retrieved
+        lazily.
+
+        cdocs, if any, should be a dictionary in which the keys are ascending
+        integers, beginning at one, and the values are dictionaries with the
+        content of the content-docs.
+        """
+        if isinstance(mdoc, SoledadDocument):
+            mdoc_id = mdoc.doc_id
+            mdoc = mdoc.content
+        else:
+            mdoc_id = None
+        if not mdoc:
+            mdoc = {}
+        self.mdoc = MetaMsgDocWrapper(doc_id=mdoc_id, **mdoc)
+
+        if isinstance(fdoc, SoledadDocument):
+            fdoc_id = fdoc.doc_id
+            fdoc = fdoc.content
+        else:
+            fdoc_id = None
+        self.fdoc = FlagsDocWrapper(doc_id=fdoc_id, **fdoc)
+        self.fdoc.set_future_doc_id(self.mdoc.fdoc)
+
+        if isinstance(hdoc, SoledadDocument):
+            hdoc_id = hdoc.doc_id
+            hdoc = hdoc.content
+        else:
+            hdoc_id = None
+        self.hdoc = HeaderDocWrapper(doc_id=hdoc_id, **hdoc)
+        self.hdoc.set_future_doc_id(self.mdoc.hdoc)
+
+        if cdocs is None:
+            cdocs = {}
+        cdocs_keys = cdocs.keys()
+        assert sorted(cdocs_keys) == range(1, len(cdocs_keys) + 1)
+        self.cdocs = dict([(key, ContentDocWrapper(**doc)) for (key, doc) in
+                           cdocs.items()])
+        for doc_id, cdoc in zip(self.mdoc.cdocs, self.cdocs.values()):
+            cdoc.set_future_doc_id(doc_id)
+
+    def create(self, store):
+        """
+        Create all the parts for this message in the store.
+        """
+        leap_assert(self.cdocs,
+                    "Need non empty cdocs to create the "
+                    "MessageWrapper documents")
+        leap_assert(self.mdoc.doc_id is None,
+                    "Cannot create: mdoc has a doc_id")
+        leap_assert(self.fdoc.doc_id is None,
+                    "Cannot create: fdoc has a doc_id")
+
+        # TODO check that the doc_ids in the mdoc are coherent
+        # TODO I think we need to tolerate the no hdoc.doc_id case, for when we
+        # are doing a copy to another mailbox.
+        # leap_assert(self.hdoc.doc_id is None,
+        # "Cannot create: hdoc has a doc_id")
```

Comment:
I don't get this second TODO. The code is tolerating the case of no hdoc.doc_id right now, isn't it? Can this TODO and the commented code be removed? Or does the TODO need more explanation?